### ISSUE TITLE
`PYTORCH_DEBUG_MODE`, better invalid index embedding lookup error message on cuda

### ISSUE THREAD

**Issue by stas00** [pytorch/pytorch#93880]
### 🚀 The feature, motivation and pitch

You can read the detailed presentation of the problem and the ensuing discussion here:
https://pytorch.slack.com/archives/C3PDTEV8E/p1675216222317729

the outcome of discussion with @ngimel and @Chillee is this:

1. introduce a new `PYTORCH_DEBUG_MODE` env var which allows for slow validations which we don't want in the normal mode
2. rewrite `nn.Embedding.__call__` to validate that the inputs aren't larger than the target matrix and provide a user-friendly error before the call is dispatched to cuda
3. bonus: while at it make the `IndexError: index out of range in self` on cpu more user friendly to align with other pytorch ops e.g.:

```
e = torch.nn.Embedding(10,10)
e(torch.tensor([[10]]))
before: IndexError: index out of range in self
after: IndexError: index 10 is out of max range 9 in self
```

**Comment by edward-io** [comment]
Somewhat related: https://github.com/pytorch/pytorch/issues/55655 proposed by @ezyang

**Comment by malfet** [comment]
It would be nice to express it as python context manager (in addition to environment variable) as later are IMO hard to document.

**Comment by stas00** [comment]
We need to make an assumption that the user doesn't know where the problem is. Unless you're suggesting a context manager around `main()`?

**Comment by malfet** [comment]
@ngimel we should probably do it via tensor modes (which could be activated using environment variable), so that community can extend initial implementation to more operators, but initial all indexing ops should be covered.

**Comment by Chillee** [comment]
@malfet Yeah, I think it's best to make it a TorchDispatchMode with some extensible options (probably... "checks_to_run" and "callback_on_failure").

**Comment by kshitij12345** [comment]
I would be interested in taking this one. @malfet @Chillee would it be ok to tag you as the reviewer?

**Comment by Chillee** [comment]
@kshitij12345 Sounds good - I was planning on working on it eventually but if you're interested even better :)

**Comment by zou3519** [comment]
Is the implication that the general approach we should be taking for CUDA asserts is to have a PYTORCH_DEBUG_MODE check that actually does some slower asserts that have better error messages?

**Comment by ngimel** [comment]
CUDA asserts are necessary for correctness, so we are not taking them out, but we want to add a slow debug mode that could provide better user debugging experience.

**Comment by albanD** [comment]
I would vote for it to imply `CUDA_LAUNCH_BLOCKING=1`!

**Comment by ngimel** [comment]
tbh, I'm not sure CUDA_LAUNCH_BLOCKING is needed/desirable - the whole point is that checks currently happening on device will be moved to CPU and will raise an error even before problematic kernel is scheduled (they also by necessity will cause device-host synchronization). If for some reason CUDA_LAUNCH_BLOCKING is indeed needed even on top of these extra synchronizations, it's easy to do on the user side.

**Comment by nkalpakis21** [comment]
i see this has been open for awhile. is there trouble getting a dev to fix this issue? what is blocking this for so long
